falcon/falcon.py

```diff
@@ -246,231 +293,266 @@ def main(args: Union[str, List[str]] = None) -> int:
+            pa.field("retention_time", pa.float32()),
+            pa.field("filename", pa.string()),
+            pa.field("vector", pa.list_(pa.float32())),
+        ]
+    )
+    lance_writers = multiprocessing.pool.ThreadPool(
+        max_file_workers,
+        _write_spectra_lance,
+        (
+            spectra_queue,
+            lance_locks,
+            schema,
+            charges,
+            vectorize,
+        ),
+    )
     # Add sentinels to indicate stopping. This needs to happen after all files
     # have been read (by joining `peak_readers`).
     for _ in range(max_file_workers):
-        spectra_queue.put((None, None))
-    pkl_writers.close()
-    pkl_writers.join()
-    # Flush all remaining spectra to pickle files.
-    for pkl_filename, pkl_spectra in spectra_to_pkl.items():
-        with open(pkl_filename, 'ab+') as f_out:
-            for spec in pkl_spectra:
-                pickle.dump(spec, f_out, protocol=pickle.HIGHEST_PROTOCOL)
-    # Make sure the spectra in the individual files are sorted by their
-    # precursor m/z and count the number of spectra per precursor charge.
-    buckets: Dict = collections.defaultdict(lambda: [0, []])
-    pkl_filenames = spectra_to_pkl.keys()
-    for filename, n_spectra_file in zip(
-            pkl_filenames, joblib.Parallel(n_jobs=-1, backend='threading')(
-                joblib.delayed(_sort_spectra_pkl)(filename)
-                for filename in pkl_filenames)):
-        charge = os.path.basename(filename)
-        charge = int(charge[:charge.index('_')])
-        buckets[charge][0] += n_spectra_file
-        buckets[charge][1].append(filename)
-    n_spectra, n_buckets = 0, 0
-    for charge, (n_spectra_charge, filenames) in buckets.items():
-        n_spectra += n_spectra_charge
-        n_buckets += len(filenames)
-        buckets[charge] = n_spectra_charge, natsort.natsorted(filenames)
-    logger.debug('%d spectra written to %d buckets by precursor charge and '
-                 'precursor m/z', n_spectra, n_buckets)
-    return dict(buckets)
-
-
-def _read_spectra_queue(file_queue: queue.Queue, spectra_queue: queue.Queue) \
-        -> None:
+        spectra_queue.put(None)
+    lance_writers.close()
+    lance_writers.join()
+
+    # Count the total number of spectra in the datasets.
+    dataset_paths = [
+        os.path.join(
+            config.work_dir, "spectra", f"spectra_charge_{charge}.lance"
+        )
+        for charge in charges
+    ]
+    n_spectra = 0
+    for dataset_path in dataset_paths:
+        try:
+            dataset = lance.dataset(dataset_path)
+        except ValueError:
+            charge = int(dataset_path.split("_")[-1].split(".")[0])
+            logger.error("Failed to create dataset for charge %d", charge)
+            charges.remove(charge)
+            continue
+        n_spectra += dataset.count_rows()
+    logger.debug(
+        "Read %d spectra from %d peak files", n_spectra, len(input_filenames)
```

Comment:
**suggestion:** This might be upgraded to the `info` level because it is useful information for the user.